# Main Assignment: 

Create a class called ShoppingCart.  Create a constructor that takes no arguments and sets the total attribute to zero, and initializes an empty dict attribute named items.

Create a method add_item that requires item_name, quantity and price arguments. This method should add the cost of the added items to the current value of total. It should also add an entry to the items dict such that the key is the item_name and the value is the quantity of the item.

Create a method remove_item that requires similar arguments as add_item. It should remove items that have been added to the shopping cart and are not required. This method should deduct the cost of the removed items from the current total and also update the items dict accordingly.

If the quantity of an item to be removed exceeds the current quantity of that item in the cart, assume that all entries of that item are to be removed.

Create a method checkout that takes in cash_paid and returns the value of balance from the payment. If cash_paid is not enough to cover the total, return "Cash paid not enough".

Create a class called Shop that has a constructor which takes no arguments and initializes an attribute called quantity at 100.

# Optional (more advanced OOP) 

Make sure Shop inherits from ShoppingCart.

In the Shop class, override the remove_item method, such that calling Shop's remove_item with no arguments decrements quantity by one.

# Selenium Web Driver (Web Dev / UI Embellishment) 

In a browser, navigation to http://automationpractice.com/ website (ensure that no extensions block the website, i.e. Norton Security). The exercise will test both selenium coding and comparison between the dresses total price and the calculated total price. 

In detail, we verify the following: 

1) All 5 dresses are listed on the summary page 
2) The total price of each dress is correct. 
3) The total products price is the sume of the total price of each dress. 

Requirements: 

Test case must be re-runnable. 

Should work on all browsers (mostly Chrome and Firefox)

Not run with any errors or browser websites incompatible 

In [3]:
class ShoppingCart():
 
    # function to initialize the shopping cart. 
    def __init__(self):
        self.dict_items = {}
        self.total = 0
    
    # function to add cost of added items 
    def add_item(self, item_name, quantity, price):
        # get the total 
        self.total = price * quantity 
        # add entry to the dict (using Python key, value shortcut)
        self.dict_items.update({item_name: quantity})
    
    # function to remove cost of removed items 
    def remove_item(self, item_name, quantity, price):
        if item_name in self.dict_items:
            # ensure no duplicates in the shopping cart for items 
            if quantity > 0 and quantity < self.dict_items[item_name]:
                # remove the item here. 
                self.dict_items[item_name] -= quantity 
                # reduce quantity from the total 
                self.total -= price * quantity 
            
            # quantity of an item to be removed exceeds the current quantity 
            # of that item in the cart, assume that all entries of that item are to be removed.
        elif quantity >= self.dict_items[item_name]:
            self.total -= self.dict_items[item_name] * price 
            # remove with "del" function
            del self.dict_items[item_name]
        
        
    # function to return the value of balance from the payment. 
    def checkout(self, cash_paid):
        # find the value of balance from payment. 
        if cash_paid < self.total:
            return "Cash paid not enough"
        balance =  cash_paid - self.total
        print("The balance is: " + str(balance))
        return balance 
        
# Shop inheritance over shopping cart class 
"""
class Shop(ShoppingCart):
    
    def __init__():
        self.quantity = 100
    
    def remove_item(self):
        self.quantity -= 1
"""

'\nclass Shop(ShoppingCart):\n    \n    def __init__():\n        self.quantity = 100\n    \n    def remove_item(self):\n        self.quantity -= 1\n'

In [4]:
# PDP style tests. 

sc = ShoppingCart()
sc.add_item("Curry 6",1,50)
sc.checkout(100)

The balance is: 50


50

In [5]:
sc.add_item("Printed Dress",2,26)
sc.add_item("Printed Summer Dress",2,29)
sc.add_item("Printed Chiffon Dress",1,16)
sc.remove_item("Printed Summer Dress",1,29)
print("The total is: " + str(sc.checkout(155) - 29))
sc.checkout(126)


The balance is: 168
The total is: 139
The balance is: 139


139

In [ ]:
def test_checkout():
    self.cart.checkout(0)

In [ ]:
# Selenium update: read from website and try to web scrape. 
# Credit: CS 506 Prof. Smith code 

from selenium import webdriver 
from time import sleep 

In [ ]:
# use selenium to open browser and XPath

driver = webdriver.Chrome("/usr/local/bin/chromedriver")
url = "http://automationpractice.com/index.php"

driver.get(url)



In [ ]:
# Read item and price in dress page 

dress_items = []

product_count = driver.find_elements_by_xpath('//*[@id="center_column"]/div[3]/div[2]/div[2]')

for i in range(len(dress_items)):
    dress_items[i] = ""
    dress_items[i]["src"] = '//*[@id="center_column"]/ul/li[1]/div/div[1]/div/a[1]/img'
    
    
    price_dress = driver.find_elements_by_xpath('//*[@id="center_column"]/ul/li[1]/div/div[2]/div[1]/span')
    dress_items[i]["price"] = prices_dress.getAttribute()
    print("Dress item: " + str(dress_items[i]["price"]))
    



In [ ]:
dict_dresses[0] = driver.find_elements_by_xpath('//*[@id="center_column"]/ul/li[1]/div/div[1]/div/a[1]/img')

    

dresses = driver.find_elements_by_xpath('//*[@id="block_top_menu"]/ul/li[2]/a')

dresses.click()

In [ ]:
# compare lists 
list_compare = []
list_actual = []

def compare_lists(list_compare, list_actual):
    for i in range(len(list_compare)):
        if list_compare[i] in list_actual[i]:
            return True 
        else: 
            return False 

# Assignment 2: HCI / Design Extension: 

Goals with the HCI Design Extension: 

1) Enhance understanding of problem space within a design

2) Provide a clear example of interaction style and decisions about conceptual design,
mainly what is the most efficient way to get the aggregate dress prices in the website and using automation techniques, compare it with the hand-calculated aggregate price?

3) Real-world application (plus potential internship / co-op task) to interactive systems and machine learning approaches to testing and training data to calculate dress pricing. Mainly, regression and classification learning.

4) Cool, in that you can build a web app whilst learning HCI :-)

The “Selenium Web Driver” application should be able to do the following:

1) Calculate all of the prices aggregate for the 5 dresses.

2) Use the XPath to get all of the necessary information.

3) Ensure that the Selenium web driver does not result in any HTTP code errors.

4) Use basics of MATLAB and data analytics to determine which customer had the best cognitive design and reasoning and the mechanisms in which HCI programs collaborate to design a solution to a complicated (on paper) problem.


# Rewriting the Shopping Cart to a Dynamic Programming Solution 

In [6]:
class ShoppingCart():
    # an array to store the items 
    array = []
    
    # function to initialize the shopping cart. 
    def __init__(self):
        self.dict_items = {}
        self.total = 0
    
    # function to add cost of added items 
    def add_item(self, item_name, quantity, price):
        # get the total 
        self.total = price * quantity 
        # add entry to the dict (using Python key, value shortcut)
        self.dict_items.update({item_name: quantity})
    
    # function to remove cost of removed items 
    def remove_item(self, item_name, quantity, price):
        # unique quantifier to avoid duplicates 
        if item_name in self.dict_items and item_name in set(self.dict_items):
            # ensure no duplicates in the shopping cart for items 
            if quantity > 0 and quantity < self.dict_items[item_name]:
                # remove the item here. 
                self.dict_items[item_name] -= quantity 
                # reduce quantity from the total 
                self.total -= price * quantity 
                # add to the array 
                array.insert(self.total)
            
            # quantity of an item to be removed exceeds the current quantity 
            # of that item in the cart, assume that all entries of that item are to be removed.
        elif quantity >= self.dict_items[item_name]:
            self.total -= self.dict_items[item_name] * price 
            # remove with "del" function
            del self.dict_items[item_name]
            # remove from the array 
            array.delete(self.total)
        
        
    # function to return the value of balance from the payment. 
    def checkout(self, cash_paid):
        # find the value of balance from payment. 
        if cash_paid < self.total:
            return "Cash paid not enough"
        balance =  cash_paid - self.total
        print("The balance is: " + str(balance))
        # memoize in the array 
        array.insert(balance)
        return balance 
        
# Shop inheritance over shopping cart class 
class Shop(ShoppingCart):
    
    def __init__():
        self.quantity = 100
    
    def remove_item(self):
        self.quantity -= 1

# Customer Class 

In [7]:
# Coding: Customer Class

# itertools import: for dictionary slicing 
from itertools import islice 

class Customer():
 
    # function to init the customer object. 
    def __init__(self):
        """Fill in here."""
        self.dict_items = {}
        self.total = 0
        self.name = ""
        self.age = 0
        self.gender = ""
        self.dresses_bought = 0
    
    # function determine total items bought 
    def buy_item(self, item_name, quantity, price):
        """Fill in here."""
        # get the total 
        self.total = price * quantity 
        # add entry to the dict (using Python key, value shortcut)
        self.dict_items.update({item_name: quantity})
        # if the customer has spent excess amount on the budget, then 
        # print out the total items that caused the excess 
        if (self.total) > 500:
            # the excess that led to the budget overspending.
            n = len(dict_items) - quantity 
            print("Items in excess")
            print(list(islice(self.dict_items,n)))
    
    # function to remove items left in the customer's budget.
    def remove_item(self, item_name, quantity, price):
        """Fill in here."""
        if item_name in self.dict_items:
            # ensure no duplicates in the shopping cart for items 
            if quantity > 0 and quantity < self.dict_items[item_name]:
                # remove the item here. 
                self.dict_items[item_name] -= quantity 
                # reduce quantity from the total 
                self.total -= price * quantity 
            
            # quantity of an item to be removed exceeds the current quantity 
            # of that item in the cart, assume that all entries of that item are to be removed.
        elif quantity >= self.dict_items[item_name]:
            self.total -= self.dict_items[item_name] * price 
            # remove with "del" function
            del self.dict_items[item_name]
            
        self.name = input("Name: ")
        if self.name == "":
            print("Customer not available, please wait.")
        # determine how many items are left 
        print("The number of items left for Customer " 
                  + self.name + " : " + str(len(self.dict_items)))
        
        
    # function confirm the checkout of the items bought from customer.
    def confirm_checkout(self, items, checkout, total_price, cash_paid):
        """Fill in here."""
        # find the value of balance from payment. 
        if cash_paid < self.total:
            return "Cash paid not enough"
        balance =  cash_paid - self.total
        print("The balance is: " + str(balance))
        
        # check the calculated price from Cashier Class, which 
        # is already defined as self.total 
        if total_price == cash_paid:
            print("Checkout approved, have a nice day.")
            self.checkout = True 
            checkout = self.checkout 
            return checkout
        self.checkout = False 
        checkout = self.checkout
        return checkout
        return balance 
        
        
    # function to have customer ask question(s), memoize the answers,
    # and get answer from cashier (similar to an actual checkout situation).
    def ask_question(self, question, answer):
        """Fill in here."""
        question = input("What is your question?")
        if question is None or question == "":
            print("Cashier not available, ask later.")
        else:
            answer = input("The answer is: ")
            print(question)
            print(answer)
        
# Shop inheritance over customer class. 
class Shop(Customer):
    
    def __init__():
        self.quantity = 100
    
    def remove_item(self):
        self.quantity -= 1
        
    def default_question(self):
        self.question = "Hello. How are you?" 

In [21]:
# Assertion Tests (PDP style)

"""Add in test cases here."""

# Test 1: customer with no items, just questions. 
customer = Customer()
customer.buy_item("Curry 6",1,50)
# good sanity check, have to add items before removing items 
# customer.remove_item("Lebron Sneakers",2,25)
# now remove proper item 
customer.remove_item("Curry 6",1,50)
customer.ask_question("", "Cashier not available, ask later.")
items = {"Curry 6": 50, "Canada Goose Coat": 1000, "Polo Shirt": 35}
customer.confirm_checkout(items, True, 1085, 1085)
# customer.confirm_checkout(items, False, 2000, 1085)

Name: Jason 
The number of items left for Customer Jason  : 1
What is your question?How to check for register price?
The answer is: On the receipt. Thank you. 
How to check for register price?
On the receipt. Thank you. 
The balance is: 1035
Checkout approved, have a nice day.


True

# Cashier Class

In [18]:
# Coding: Cashier Class 

class Cashier(): 
    
    # function to init the cashier (assume one cashier per set of customers.)
    def __init__(self):
        """Fill in here."""
        self.dict_items = {}
        self.total = 0
        self.name = ""
        self.age = 0
        self.gender = ""
        self.dresses_bought = 0
        # Set false whenever cashier closes register.
        self.open = True 
        self.checkout = False 
        
    # function for cashier to determine if the items checked out
    # is valid in value and in quantity. 
    def checkout_item(self, items, checkout, total_price, quantity, valid):
        """Fill in here."""
        # check validity of the items. 
        cashier_total = 0
        item_name = ""
        for i in range(len(items)):
            item_name = input("Enter item name: ")
            i += 1 
        total_price = 0
        for i in range(len(items)):
            total_price = eval(input())
        # register dictionary to store the items. 
        register = {}
         # get the total 
        self.total = total_price * quantity 
        # add entry to the dict (using Python key, value shortcut)
        register.update({item_name: quantity})
        self.dict_items.update({item_name: quantity})
        # if the customer has spent excess amount on the budget, then 
        # print out the total items that caused the excess 
        if len(self.dict_items) != len(register):
            print("Checked items not valid.")
            return self.checkout 
        if len(self.dict_items) == 0 or len(register) == 0:
            print("Checked items not valid.")
            return self.checkout
        else:
            self.checkout = True
            valid = True 
            print("Checked out items valid.")
            # print out the key value pairs for the items, as a receipt
            print("Your receipt is: \n")
            for key, value in self.dict_items.items() :
                print(key, value)
           
        
    # function for cashier to answer a customer's question.
    def answer_question(self, answer, question):
        """Fill in here."""
        name = input()
        # ensure customer has no more items to check out 
        self.checkout = False 
        if question is None or question == "" or len(question) == 0:
            print("Did not get the question. Try again.")
        cashier_name = "Cashier: " + name 
        answer = input("The answer is: ")
        print(question)
        print(answer)
        print(cashier_name + " has checked out your items. Have a good day.")
        
# Shop inheritance over customer class. 
class Shop(Cashier):
    
    def __init__():
        self.quantity = 100
    
    def remove_item(self):
        self.quantity -= 1
        
    def default_question(self):
        self.question = "Hello. How are you?" 

In [19]:
# Assertion Tests (PDP style)

"""Add in test cases here. """

# Cashier 
cash = Cashier() 
items = {"Curry 6": 50, "Canada Goose Coat": 1000, "Polo Shirt": 35}
# sum the item values. 
verify_checkout = sum(items.values())
print("Final total: " + str(verify_checkout))
cash.checkout_item(items, True, 1085, 1085, True)
cash.answer_question("On the register", "How may I verify the checkout?")

Final total: 1085
Enter item name: Curry 6
Enter item name: Canada Goose Coat 
Enter item name: Polo Shirt 
50
1000
35
Checked out items valid.
Your receipt is: 

Polo Shirt  1085
Dora
The answer is: On the register 
How may I verify the checkout?
On the register 
Cashier: Dora has checked out your items. Have a good day.


# The customer interaction: 

This is just an amalgamation of a potential customer interaction between the cashier, customer: 
Bold: denotes the verbatim output. 

Obviously, given that we are using strict HCI principles of cognition, the results are rather robotic, and not very human. The program just simulates what was recalled in the interaction between cashier and customer. 

Hello Jason, here is your balance from Selenium Shops: 

#### The balance is: 168
#### The total is: 139
#### The balance is: 139
Total balance is: 
#### 139

Your interaction was described as follows: 

#### Name: Jason 
Items checked out: 

Curry 6               50 dollars 
Canada Goose Coat:  1000 dollars 
Polo Shirts:          35 dollars 
#### The number of items left for Customer Jason  : 1
#### What is your question?How may I check the items left? 
#### The answer is: You can check in your receipt from the cashier. 

Asked Cashier Dora: 

#### How may I check the items left? 
You can check in your receipt from the cashier. 
The balance is: 1035
Checkout approved, have a nice day.
The balance is: 1034
False

Answered by Cashier Dora: 
#### Final total: 1085
#### Enter item name: Curry 6
#### Enter item name: Canada Goose Coat 
#### Enter item name: Polo Shirt 
#### 50
#### 1000
#### 35
#### Checked out items valid.

#### Your receipt is: 

Selenium Shops Inc. 
_______________________

Curry 6 Sneakers  $50

Canada Goose Coat: $1000

Polo Shirt: $35 

Total: $1085 

#### The answer is: On the register 
#### How may I verify the checkout?
#### On the register 
#### Cashier Dora has checked out your items. Have a good day.


# Answer to Q3 (Dictionary vs. Hash Table vs. List Tuple For Cognition) 

The reason why we use a dictionary as compared to the hash table or list tuple is mainly for ability to make immutable cognition processes, or in other words, once something has been confirmed, we cannot take it back. In addition, the dictionary <key,value> property in Python allows us to quickly verify if a particular item has been misplaced by a pricing, or vice a versa. Each key in a dictionary can only be paired with one unique value, which makes the dictionary as a data structure immutable without re-writing the values to the array, much like the register and receipt in a real world shopping scenario. 

If we had to use a hash table, we would have to know what was the hashing value of the key to the value (as in a cryptographic fashion), and for a real-life shopping application, that would simply be impractical. Also, for the list tuple, we can only hold a finite amount of items and that it would be very unrealistic to know exactly what index the item was put into, let alone the key and value. 

So, in summary, the dictionary is preferable for cognition and information processes for HCI applications, such as this Selenium Shops Python Application, for the following: 

#### Runtime:

Dictionary: <key,value> search: O(1) 

Hash Table: O(1) without hash, O(n) with hash 

List Tuple: O(n^2) with full list, O(n) if just linear search 

#### Practicality: 

Dictionary: Can be practical, "human friendly", similar to that of a human interaction between checking the item names and matching them with item prices on a real shopping list. 

Hash Table: Not practical for cryptographic reasons, not every hash can be to each item in the table, and linear collision would mean the non CS equivalent of having duplicate items. Selenium Shops, which mirrors a big outlet retailer like Macys, Premium Outlets, seldom make that type of mistake, or the cashier would be fired. 

List Tuple: Simply too much to memorize in a system, a key and value pair to each index of the tuple. 

#### Usability: 

Dictionary: Usable for any finite number of items, even if the receipt had missing items (a real life possibility,
considering human error in checkouts).  

Hash Table: Useful until duplicate items or missing items occur. 

List Tuple: Simply not practical at all for application. 

# Summary of the Chasier, Customer, and Shopping Cart Interaction: 

# Coding From Matlab to draw the line graph for each customer purchase in the interaction. 

CreateLinePlotFromMatrixExample.mlx 

% HCI Pre Lecture Assignment 2 
% Coding for the line graph of each customer's purchase 

% Items purchased by customer Jason 
% x-axis: the number of items purchased 
$ y-axis: the price of each item purchased 

jason = [1 2 3 2 2 1 1; 50 1000 35 26 29 16 29]

% Plot the Jason Graph. 

figure
plot(jason) 

% scaling for the graph.
xlim([1.000, 2.000])
ylim([1 1000])

% Add some labels for the graph. 
title('Selenium Shops Purchases From Customer Jason')
% add the x-axis and y-axis captions 
xlabel('Number of items purchased') 
ylabel('The price of each item (in $)') 

# Resulting Line Graph From Code: 

![alt text](line_graph_matlab.png "Selenium Shops Line Graph")